File: reverse_linked_list.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

def reverse(previous, current):
    if not current.next:
        current.next = previous
        return current
    
    new_head = reverse(current, current.next)
    current.next = previous

    return new_head



if __name__ == '__main__':
    head = ListNode(1)
    head.next = ListNode(2)
    head.next.next = ListNode(3)
    head.next.next.next = ListNode(4)

    new_head = reverse(None, head)
    print(new_head.val)
    print(new_head.next.val)
    print(new_head.next.next.val)
    print(new_head.next.next.next.val)

File: problems/reverse_linked_list.rb
class ListNode
  attr_accessor :val, :next

  def initialize(x)
    @val = x
    @next = nil
  end
end

def reverse_list(previous, current)
  if current.next.nil?
    current.next = previous
    return current
  end

  new_head = reverse_list(current, current.next)
  current.next = previous

  return new_head
end

head = ListNode.new(1)
head.next = ListNode.new(2)
head.next.next = ListNode.new(3)
head.next.next.next = ListNode.new(4)
new_head = reverse_list(nil, head)
p new_head.val
p new_head.next.val
p new_head.next.next.val
p new_head.next.next.next.val
